marketo-client/client.go

```diff
@@ -0,0 +1,260 @@
+			},
+		},
+		"fields": fields,
+	})
+	if err != nil {
+		return "", err
+	}
+	path := "/bulk/v1/leads/export/create.json"
+	response, err := c.Post(path, reqBody)
+	if err != nil {
+		return "", err
+	}
+	if !response.Success {
+		return "", fmt.Errorf("%+v", response.Errors)
+	}
+	var result []CreateExportResult
+	if err := json.Unmarshal(response.Result, &result); err != nil {
+		return "", err
+	}
+	if len(result) != 1 {
+		return "", fmt.Errorf("unexpected response from marketo rest api:%+v", result)
+	}
+	return result[0].ExportID, nil
+}
+
+type CreateExportResult struct {
+	ExportID  string    `json:"exportId"`
+	Format    string    `json:"format"`
+	Status    string    `json:"status"`
+	CreatedAt time.Time `json:"createdAt"`
+}
+
+func (c Client) EnqueueExportLeads(exportID string) (string, error) {
+	path := fmt.Sprintf("/bulk/v1/leads/export/%s/enqueue.json", exportID)
+	response, err := c.Post(path, nil)
+	if err != nil {
+		return "", err
+	}
+
+	if !response.Success {
+		if response.Errors[0].Code == "1029" {
+			return "", ErrEnqueueLimit
+		}
+		return "", fmt.Errorf("%+v", response.Errors)
+	}
+	return exportID, nil
+}
+
+func (c Client) StatusOfExportLeads(exportID string) (StatusOfExportResult, error) {
+	path := fmt.Sprintf("/bulk/v1/leads/export/%s/status.json", exportID)
+	response, err := c.Get(path)
+	if err != nil {
+		return StatusOfExportResult{}, err
+	}
+	if !response.Success {
+		return StatusOfExportResult{}, fmt.Errorf("%+v", response.Errors)
+	}
+	var result []StatusOfExportResult
+	if err := json.Unmarshal(response.Result, &result); err != nil {
+		return StatusOfExportResult{}, err
+	}
+	if len(result) != 1 {
+		return StatusOfExportResult{}, fmt.Errorf("unexpected response from marketo rest api:%+v", result)
+	}
+	return result[0], nil
+}
+
+type StatusOfExportResult struct {
+	ExportID        string    `json:"exportId"`
+	Format          string    `json:"format"`
+	Status          string    `json:"status"`
+	CreatedAt       time.Time `json:"createdAt"`
+	QueuedAt        time.Time `json:"queuedAt"`
+	StartedAt       time.Time `json:"startedAt"`
+	FinishedAt      time.Time `json:"finishedAt"`
+	NumberOfRecords int       `json:"numberOfRecords"`
+	FileSize        int       `json:"fileSize"`
+	FileChecksum    string    `json:"fileChecksum"`
+}
+
+func (c Client) CancelExportLeads(exportID string) error {
+	path := fmt.Sprintf("/bulk/v1/leads/export/%s/cancel.json", exportID)
+	response, err := c.Post(path, nil)
+	if err != nil {
+		return err
+	}
+	if !response.Success {
+		if response.Errors[0].Code == "1029" {
+			return ErrCannotCancel
+		}
+		return fmt.Errorf("%+v", response.Errors)
+	}
+	return nil
+}
+
+func (c Client) FileExportLeads(ctx context.Context, endpoint string, exportID string) (*[]byte, error) {
+	path := fmt.Sprintf("/bulk/v1/leads/export/%s/file.json", exportID)
+
+	req, err := http.NewRequestWithContext(ctx, "GET", endpoint+path, nil)
+	if err != nil {
+		return nil, err
```

Comment:
We should wrap `err` and add more context about it.

Reply:
updated